### PR TITLE
Bump PureScript to 0.14.3

### PR DESCRIPTION
**Pull request by ptrfrncsmrph**
All told, only ~17~ 14 of 80 ended up in the `broken` pile which doesn't seem too bad.

---

- [x] AceEditorHalogenHooks
  :x: Move to `broken`
  ```
  Module Halogen.Query.EventSource was not found.
  ```
- [x] Add RemoveEventListenerJs
  :pencil2: Add missing dependencies to `spago.dhall`
- [x] AffGameSnakeJs
  :x: Move to `broken`
  ```
  [error] The following packages do not exist in your package set:
  [error]   - canvas-action
  [error]   - game
  ```
- [x] AffjaxPostNode
  :pencil2: Add missing dependencies to `spago.dhall`
- [x] BasicHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] BehaviorSuperCircleJs
  :x: Move to `broken`
  ```
  [error] The following packages do not exist in your package set:
  [error]   - behaviors
  ```
- [x] BigIntJs
  :white_check_mark: Already works
- [x] BookHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
- [x] BookReactHooks
  :white_check_mark: Already works
- [x] ButtonsHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] ButtonsReactHooks
  :white_check_mark: Already works
- [x] CapabilityPatternNode
  :white_check_mark: Already works
- [x] CardsHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
  :pencil2: Change `NonEmpty Array` to `NonEmptyArray`
- [x] CardsReactHooks
  :pencil2: Change `NonEmpty Array` to `NonEmptyArray`
- [x] CatGifsHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] CatGifsReactHooks
  :white_check_mark: Already works
- [x] ClockReactHooks
  :white_check_mark: Already works
- [x] ComponentsHalogenHooks
  :x: Move to `broken`
  ```
  Cannot import type Hooked from module Halogen.Hooks
  It either does not exist or the module does not export it.
  ```
- [x] ComponentsInputHalogenHooks
  :x: Move to `broken`
  ```
  Cannot import type Hooked from module Halogen.Hooks
  It either does not exist or the module does not export it.
  ```
- [x] ComponentsInputReactHooks
  :white_check_mark: Already works
- [x] ComponentsMultiTypeHalogenHooks
  :x: Move to `broken`
  ```
  Cannot import type Hooked from module Halogen.Hooks
  It either does not exist or the module does not export it.
  ```
- [x] ComponentsMultiTypeReactHooks
  :white_check_mark: Already works
- [x] ComponentsReactHooks
  :white_check_mark: Already works
- [x] DateTimeBasicsLog
  :white_check_mark: Already works
- [x] DebuggingLog
  :pencil2: Change `Debug.Trace` to `Debug`
- [x] DiceCLI
  :pencil2: Switch argument order
- [x] DiceLog
  :white_check_mark: Already works
- [x] DragAndDropHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] DragAndDropReactHooks
  :white_check_mark: Already works
- [x] DriverIoHalogenHooks
  :x: Move to `broken`
  ```
  Module Control.Coroutine was not found.
  ```
- [x] DriverRoutingHalogenHooks
  :x: Move to `broken`
  ```
  Could not match type

    Function t1

  with type

    Proxy @Symbol
  ```
- [x] DriverWebSocketsHalogenHooks
  :x: Move to `broken`
  ```
  Could not match type

    Function t1

  with type

    Proxy @Symbol
  ```
- [x] FileUploadHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] FileUploadReactHooks
  :white_check_mark: Already works
- [x] FindDomElementJs
  :white_check_mark: Already works
- [x] FormsReactHooks
  :white_check_mark: Already works
- [x] GroceriesHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
- [x] GroceriesJs
  :pencil2: `appendChild` returns `Unit` instead of `Node`
- [x] GroceriesReactHooks
  :white_check_mark: Already works
- [x] HelloConcur
  :white_check_mark: Already works
- [x] HelloHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
- [x] HelloJs
  :white_check_mark: Already works
- [x] HelloLog
  :white_check_mark: Already works
- [x] HelloReactHooks
  :white_check_mark: Already works
- [x] HeterogenousArrayLog
  :pencil2: Replace `SProxy` with `Proxy`
- [x] ImagePreviewsHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] ImagePreviewsReactHooks
  :white_check_mark: Already works
- [x] InterpretHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] KeyboardInputHalogenHooks
  :x: Move to `broken`
  ```
  Module Halogen.Query.EventSource was not found.
  ```
- [x] LifecycleHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
  :pencil2: Replace `SProxy` with `Proxy`
- [x] MemoizeFibonacci
  :x: Move to `broken`
  ```
  [error] The following packages do not exist in your package set:
  [error]   - memoize
  ```
- [x] NumbersHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] NumbersReactHooks
  :white_check_mark: Already works
- [x] ParallelAppMExampleLog
  :white_check_mark: Already works
- [x] PayloadHttpApiNode
  :white_check_mark: Already works
- [x] PositionsHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] PositionsReactHooks
  :white_check_mark: Already works
- [x] RandomNumberGameNode
  :white_check_mark: Already works
- [x] ReadPrintFileContentsConcur
  :white_check_mark: Already works
- [x] ReadPrintFileContentsNode
  :white_check_mark: Already works
- [x] RoutingHashHalogenClassic
  :x: Move to `broken`
  ```
  Could not match type

    Function t1

  with type

    Proxy @Symbol
  ```
- [x] RoutingHashLog
  :pencil2: Remove `generics-rep` from `spago.dhall`
  :pencil2: Replace `Data.Generic.Rep.Show` with `Data.Show.Generic`
- [x] RoutingHashReactHooks
  :pencil2: Remove `generics-rep` from `spago.dhall`
- [x] RoutingPushHalogenClassic
  :x: Move to `broken`
  ```
  Could not match type

    Function t1

  with type

    Proxy @Symbol
  ```
- [x] RoutingPushReactHooks
  :pencil2: Remove `generics-rep` from `spago.dhall`
- [x] RunCapabilityPatternNode
  :pencil2: Update usage of `Run`
- [x] ShapesHalogenHooks
  :pencil2: Update to reflect a few breaking changes in `Halogen.Svg.Attributes`
- [x] ShapesReactHooks
  :white_check_mark: Already works
- [x] SignalRenderJs
  :white_check_mark: Already works
- [x] SignalSnakeJs
  :white_check_mark: Already works
- [x] SignalTrisJs
  :pencil2: Use `toMap` from `Data.Set`
  :pencil2: Remove `toNonEmpty`
- [x] SimpleASTParserLog
  :pencil2: Replace `Global` import with qualified `Data.Number` import
- [x] TextFieldsHalogenHooks
  :pencil2: Remove `HH.HTML` as parameter to `Component`
  :pencil2: Update event handler signature from `MouseEvent -> Maybe i` to `MouseEvent -> i`
- [x] TextFieldsReactHooks
  :white_check_mark: Already works
- [x] TicTacToeReactHooks
  :pencil2: Remove `generics-rep` from `spago.dhall`
  :pencil2: Replace `Data.Generic.Rep.Show` with `Data.Show.Generic`
- [x] TimeHalogenHooks
  :x: Move to `broken`
  ```
  Module Halogen.Query.EventSource was not found.
  ```
- [x] TimeReactHooks
  :white_check_mark: Already works
- [x] ValueBasedJsonCodecLog
  :pencil2: Replace partial `fromRight` with inline lambda
- [x] WindowPropertiesJs
  :white_check_mark: Already works
- [x] WriteFileNode
  :white_check_mark: Already works

### CODE REVIEW

**Comment by JordanMartinez** [comment]
Thanks for working on this!

File: broken/AceEditorHalogenHooks/src/Main.purs [hunk 1]
```diff
@@ -33,7 +33,7 @@ _ace = SProxy :: SProxy "ace"
 containerComponent
   :: forall unusedQuery unusedInput unusedOutput anyMonad
    . MonadAff anyMonad
-  => H.Component HH.HTML unusedQuery unusedInput unusedOutput anyMonad
+  => H.Component unusedQuery unusedInput unusedOutput anyMonad
```
Comment (ptrfrncsmrph) [inline broken/AceEditorHalogenHooks/src/Main.purs:36]: I ran a `sed` command to remove `HH.HTML`
```
sed -i "s/Component HH\.HTML/Component/g" ./recipes/*HalogenHooks/*/Main.purs
```
but some recipes remained broken after that and other minimal attempts at fixing, which explains why some of these `HalogenHooks` recipes have the `HH.HTML` removed but still ended up in the `broken` directory. I think all the other recipes that ended up in `broken` were left unchanged.

File: packages.dhall [hunk 2]
```diff
@@ -1,5 +1,5 @@
 let upstream =
-      https://github.com/purescript/package-sets/releases/download/psc-0.13.8-20200831/packages.dhall sha256:cdb3529cac2cd8dd780f07c80fd907d5faceae7decfcaa11a12037df68812c83
+      https://github.com/purescript/package-sets/releases/download/psc-0.14.2-20210713/packages.dhall sha256:654c3148cb995f642c73b4508d987d9896e2ad3ea1d325a1e826c034c0d3cd7b
```
Comment (ptrfrncsmrph) [inline packages.dhall:2]: I realized halfway through that this is actually slightly ahead of TPS package set (https://github.com/purescript/trypurescript/blob/ee9ca742c6c34901303043f64815c9794d23cd4c/staging/packages.dhall#L2) I don't know that that would lead to any incompatibility

File: recipes/ValueBasedJsonCodecLog/src/Main.purs [hunk 3]
```diff
@@ -46,9 +46,8 @@ main = do
   log $ "Encoding the example value:"
   log $ stringify $ encode entireRecordCodec exampleValue
 
-
 exampleJson :: Json
-exampleJson = unsafePartial $ fromRight $ jsonParser
+exampleJson = unsafePartial $ (\(Right json) -> json) $ jsonParser
```
Comment (ptrfrncsmrph) [inline recipes/ValueBasedJsonCodecLog/src/Main.purs:50]: `fromRight` is no longer a partial function https://pursuit.purescript.org/packages/purescript-either/5.0.0/docs/Data.Either#v:fromRight

File: recipes/SimpleASTParserLog/src/Main.purs [hunk 4]
```diff
@@ -137,12 +137,12 @@ parseAtom = do
       void $ string "." -- decimal point
       decimalsAsString <- parseNumSequence
       let fullString = digitsAsString <> "." <> decimalsAsString
-      case readFloat fullString of
-        x | isFinite x -> pure $ LitNum x
+      case Number.fromString fullString of
+        Just x | Number.isFinite x -> pure $ LitNum x
```
Comment (ptrfrncsmrph) [inline recipes/SimpleASTParserLog/src/Main.purs:141]: Couldn't find the partial `readFloat` here, so replaced with `fromString`. Since there was already a `fromString` imported from `Data.Int`, I had to namespace both imports to disambiguate.

Comment (JordanMartinez) [reply recipes/SimpleASTParserLog/src/Main.purs:141]: The `Number.isFinite` check is no longer necessary as `Number.fromString` [already](https://github.com/purescript/purescript-numbers/blob/v8.0.0/src/Data/Number.purs#L53) handles [that](https://github.com/purescript/purescript-numbers/blob/v8.0.0/src/Data/Number.js#L12-L19).

File: packages.dhall (hunk 2 above)
Comment (JordanMartinez) [reply packages.dhall:2]: It shouldn't, but thanks for notifying us of that.

File: recipes/SimpleASTParserLog/src/Main.purs (hunk 4 above)
Comment (ptrfrncsmrph) [reply recipes/SimpleASTParserLog/src/Main.purs]: 0668a5d

**Comment by JordanMartinez** [comment]
- Argument order was switched in [node-readline#23](https://github.com/purescript-node/purescript-node-readline/pull/23/files):
    - DiceCLI
- `FProxy` was replaced with `Proxy`
    - RunCapabilityPatternNode
- `Map`'s `Monoid` instance was dropped. Use [`SemigroupMap`](https://pursuit.purescript.org/packages/purescript-ordered-collections/2.0.1/docs/Data.Map#t:SemigroupMap) newtype to get instance again, but check the `k`'s `Monoid` instance to verify that implementation is still the same:
    - SignalTrisJs
    
Don't worry about these ones:

- Needs to update code to use new EventSource code
    - AceEditorHalogenHooks - [repo example](https://github.com/purescript-halogen/purescript-halogen/blob/master/examples/ace/src/AceComponent.purs)
    - DriverIoHalogenHooks - [repo example](https://github.com/purescript-halogen/purescript-halogen/blob/master/examples/driver-io/src/Main.purs)
    - DriverWebSocketsHalogenHooks - [repo example](https://github.com/purescript-halogen/purescript-halogen/blob/master/examples/driver-websockets/src/Main.purs)
    - KeyboardInputHalogenHooks - 
    - TimeHalogenHooks
- `Hooked` type missing from `Halogen.Hooks` module. [declaration of hook](https://github.com/JordanMartinez/purescript-halogen-hooks-extra/pull/10/files#diff-f94db1a60f245b5fb67c4c09e1c6eb0a284d4137e3ae7d8239193920d53bee56R20-R25) and [the function that creates/uses it](https://github.com/JordanMartinez/purescript-halogen-hooks-extra/pull/10/files#diff-f94db1a60f245b5fb67c4c09e1c6eb0a284d4137e3ae7d8239193920d53bee56R56-R66)
    - ComponentsHalogenHooks
    - ComponentsInputHalogenHooks
    - ComponentsMultiTypeHalogenHooks
- `slot_` was [changed](https://github.com/JordanMartinez/purescript-halogen-hooks-extra/pull/10/files#diff-1b64e0141b229279ff50387b8a0a71f1d33532d1fd4696a5ae11db88b6f20656R22-R25)
    - DriverRoutingHalogenHooks
    - RoutingHashHalogenClassic
    - RoutingPushHalogenClassic

Can't do anything about these ones other than get the packages added again to the package set:

- Packages missing from package set. 
    - canvas-action, game
        - AffGameSnakeJs
    - behaviors
        - BehaviorSuperCircleJs
    - memoize
        - MemoizeFibonacci

File: recipes/ValueBasedJsonCodecLog/src/Main.purs (hunk 3 above)
Comment (ptrfrncsmrph) [reply recipes/ValueBasedJsonCodecLog/src/Main.purs:50]: I considered making a standalone function for this, something like
```haskell
fromRight :: Partial => forall a b. Either a b -> b
fromRight (Right x) = x
```
but I opted for the minimal change. I'm definitely open to suggestions here; this one in particular felt like one of the more questionable changes.

**Comment by ptrfrncsmrph** [comment]
>     * Argument order was switched in [node-readline#23](https://github.com/purescript-node/purescript-node-readline/pull/23/files):
>       
>       * DiceCLI
> 
>     * `FProxy` was replaced with `Proxy`
>       
>       * RunCapabilityPatternNode
> 
>     * `Map`'s `Monoid` instance was dropped. Use [`SemigroupMap`](https://pursuit.purescript.org/packages/purescript-ordered-collections/2.0.1/docs/Data.Map#t:SemigroupMap) newtype to get instance again, but check the `k`'s `Monoid` instance to verify that implementation is still the same:
>       
>       * SignalTrisJs
> 

Sounds good 👍 those seem reasonable enough to tackle in this PR

**Comment by JordanMartinez** [comment]
Thanks again for all this work!

File: recipes/RunCapabilityPatternNode/src/Application.purs [hunk 5]
```diff
@@ -36,11 +36,11 @@ program = do
 derive instance loggerFunctor :: Functor LoggerF
 derive instance getUserNameFunctor :: Functor GetUserNameF
 
-type LOGGER = FProxy LoggerF
-type GET_USER_NAME = FProxy GetUserNameF
+type LOGGER r = (logger :: LoggerF | r)
+type GET_USER_NAME r = (getUserName :: GetUserNameF | r)
```
Comment (ptrfrncsmrph) [inline recipes/RunCapabilityPatternNode/src/Application.purs:40]: This compiles and works, but I was just guessing my way through this refactor while referencing the [`Run` docs](https://github.com/natefaubion/purescript-run#writing-interpreters).

File: recipes/SignalTrisJs/src/Main.purs [hunk 6]
```diff
@@ -398,11 +397,6 @@ setFillStyleC :: Context2D -> Color -> Effect Unit
 setFillStyleC ctx = setFillStyle ctx <<< toHexString
 
 -----------------
--- Faster versions of these functions are proposed in
--- https://github.com/purescript/purescript-ordered-collections/pull/21
-toMap :: forall a. Ord a => Set a -> Map a Unit
-toMap = foldl (\m k -> Map.insert k unit m) mempty
-
 annotateSet :: forall a b. Ord a => (a -> b) -> Set a -> Map a b
 annotateSet f = mapWithIndex (\k _ -> f k) <<< toMap
```
Comment (ptrfrncsmrph) [inline recipes/SignalTrisJs/src/Main.purs:401]: [`toMap` is now in `Data.Set`](https://github.com/purescript/purescript-ordered-collections/pull/31/files), so no need to implement `Monoid`. I couldn't find anything like `annotateSet` in there though.

File: package.json [hunk 7]
```diff
@@ -2,8 +2,8 @@
   "license": "MIT",
   "devDependencies": {
     "parcel-bundler": "^1.12.5",
-    "purescript": "^0.13.8",
-    "spago": "^0.16.0"
+    "purescript": "^0.14.3",
+    "spago": "^0.19.1"
```
Comment (JordanMartinez) [inline package.json:6]: We'll want to bump this to the latest spago version, but that should be done in a separate PR.

File: recipes/ValueBasedJsonCodecLog/src/Main.purs (hunk 3 above)
Comment (JordanMartinez) [reply recipes/ValueBasedJsonCodecLog/src/Main.purs:50]: I think people have typically been doing something like `either (\_ -> unsafeCrashWith "got Left: bad parser") identity`. It's a bit longer, but then it at least provides a somewhat clearer error message.

**Review by JordanMartinez — APPROVED** [review]
Changes to make before merging this:
- [ ] Let's change migrate from the `fromRight` partial function to <del>`either (\_ -> unsafeCrashWith "got Left: bad parser") identity`</del> `fromRight' (\_ -> unsafeCrashWith "got Left: bad parser")`.
- [ ] Check whether `v0.14.3` compiles this PR too (if it does, use it; if not, ignore it and continue with the one you have).

Thank you for all your work!

File: packages.dhall (hunk 2 above)
Comment (JordanMartinez) [reply packages.dhall:2]: Also, a `v0.14.3` package set was just released. Could you verify whether or not using that package set works?

**Comment by ptrfrncsmrph** [comment]
>     * [ ]  Let's change migrate from the `fromRight` partial function to `either (\_ -> unsafeCrashWith "got Left: bad parser") identity`.

I think
```haskell
either (\_ -> unsafeCrashWith "got Left: bad parser") identity
```
is same as 
```haskell
fromRight' \_ -> unsafeCrashWith "got Left: bad parser"
```
I'm fine with either (no pun intended) but would you still prefer the former?

**Comment by JordanMartinez** [comment]
Oh, whoops! Yeah, let's use the `fromRight' \_ -> ...` version.

**Comment by ptrfrncsmrph** [comment]
`make buildAll` works with the 0.14.3 package set 👍

File: recipes/SignalTrisJs/src/Main.purs (hunk 6 above)
Comment (milesfrain) [reply recipes/SignalTrisJs/src/Main.purs:401]: Just added `annotateSet` to the todo list in https://github.com/purescript/purescript-ordered-collections/issues/50

Comment (ptrfrncsmrph) [reply recipes/SignalTrisJs/src/Main.purs]: 1616b74735f1a6bc2d1950f065e366ac6bca71c9

File: recipes/ValueBasedJsonCodecLog/src/Main.purs (hunk 3 above)
Comment (ptrfrncsmrph) [reply recipes/ValueBasedJsonCodecLog/src/Main.purs]: a1f97fe

File: packages.dhall (hunk 2 above)
Comment (ptrfrncsmrph) [reply packages.dhall]: 4848fb1

**Comment by JordanMartinez** [comment]
I'm going to use a merge commit for this PR because it touches so many different recipes.
